### ISSUE TITLE
Recommend changing double quotes to single quotes wherever it is safe to do so

### ISSUE THREAD

**Issue by kurahaupo** [koalaman/shellcheck#3511]
#### For new checks and feature suggestions
- [X] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows a potential problem:

```sh
#!/bin/sh
echo "Hello World"
```

#### Here's what shellcheck currently says:

(nothing)

#### Here's what I wanted to see:

```sh
echo "Hello World"
     ^--- SC-XXXX: prefer single quotes unless quoting an expansion
Suggestion: echo 'Hello world'
```

#### Application ####

Identify all double-quoted strings, then:
* exclude strings that contain a  `'` (because we don't want to recommend rewriting `"say 'cheese'!"` as `'say '\\''cheese'\\''!')
* exclude strings that contain an expansion (in practice, just look for an unescaped <tt>\`</tt> , or an unescaped `$` that is followed by  a digit or a letter or an underscore or any of `{`, `(`, `[` , `@`, `#`, `$`, `*`, `?`).
* exclude strings that are not just one quoted segment (e.g. `"Foo Bar"` would be reported but `'Foo'\ "Bar"` would not.
* change `\"` to just `"` in any suggested replacement

#### Rationale ####

If one consistently uses single quotes whenever double quotes are not _required_, then the presence of double quotes becomes an indication to the reader that they have to pay more attention as something "special" is happening. (This is similar to the rationale for preferring `$var` over `${var}`. In both cases the _avoidance_ of "consistency" is intentional and preferable, because it provides clearer signalling.)

**Comment by brother** [comment]
The project has histrionically more leaned the other way and we have an optional check `quote-safe-variables` to suggest double quotes for all safe cases.

If at all included make it optional.

**Comment by kurahaupo** [comment]
@brother This is independent of quoting for variables. Obviously single-quoting variables does not produce the same behaviour.

And that is the point: by using single quotes when there _isn't_ a variable (or other expansion) inside, the reader knows they can skim to the end without inspecting to see if there's one embedded somewhere.
